### PR TITLE
CONSOLE-4400: Upgrade TypeScript to v5

### PR DESCRIPTION
**Pull request by logonoff**

### CODE REVIEW

**Comment by logonoff** [comment]
Code Approver:
/assign @vojtechszocs

QE Approver:
/assign @yapei

Docs Approver:
/assign @opayne1

PX Approver:
/assign @reestr

File: dynamic-demo-plugin/tsconfig.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
     "baseUrl": ".",
     "module": "esnext",
     "moduleResolution": "node",
-    "target": "es2016",
+    "target": "es2021",
```
Comment (vojtechszocs) [inline dynamic-demo-plugin/tsconfig.json:6]: @spadgett FYI - we plan to upgrade our build target to ES 2021.

We can double check this, but recent versions of all Console supported web browsers should fully support this target.

File: frontend/@types/console/index.d.ts [hunk 2]
```diff
@@ -67,24 +67,5 @@ declare interface Window {
   Cypress?: {};
 }
 
-// TODO: Remove when upgrading to TypeScript 4.1.2+, which has a type for ListFormat and RelativeTimeFormat.
```
Comment (vojtechszocs) [inline frontend/@types/console/index.d.ts:70]: :+1:

File: frontend/packages/console-dynamic-plugin-sdk/scripts/utils/typescript.ts [hunk 3]
```diff
@@ -22,16 +22,21 @@ export const getSchemaGeneratorConfig = (srcFile: string, typeName?: string): ts
...
   }
 
-  const hasDiagnosticErrors = diagnostics.some((d) => d.category === ts.DiagnosticCategory.Error);
-  if (hasDiagnosticErrors) {
-    throw new Error(`Detected errors while parsing ${relativePath(config.path)}`);
+  const diagnosticsErrors = diagnostics.filter((d) => d.category === ts.DiagnosticCategory.Error);
+
+  if (diagnosticsErrors.length > 0) {
+    throw new Error(
+      `Detected errors while parsing ${relativePath(config.path ?? 'unknown')}, ${String(
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/scripts/utils/typescript.ts:36]: Just wondering, why do we need the `?? 'unknown'` fallback?

Comment (logonoff) [reply frontend/packages/console-dynamic-plugin-sdk/scripts/utils/typescript.ts:36]: The type for `config.path` is actually `string | undefined` but `relativePath` cannot except `undefined`

File: frontend/packages/console-dynamic-plugin-sdk/package.json [hunk 4]
```diff
@@ -22,7 +22,7 @@
     "@types/fs-extra": "9.x",
     "ejs": "3.x",
     "fs-extra": "9.x",
-    "ts-json-schema-generator": "0.98.0",
+    "ts-json-schema-generator": "2.3.0",
```
[the remark this answers is not in the document]
Comment (logonoff) [reply frontend/packages/console-dynamic-plugin-sdk/package.json:25]: I just ran `diff` on this PR vs `master` and there seems to be no changes

File: dynamic-demo-plugin/tsconfig.json (hunk 1 above)
Comment (spadgett) [reply dynamic-demo-plugin/tsconfig.json:6]: Yes, absolutely. Let's update this. I confirmed our supported browsers will all work with es2021.

**Review by spadgett — APPROVED** [review]
/approve

Great work @logonoff

**Comment by logonoff** [comment]
tech debt, adding labels

/unassign opayne1
/unassign reestr

/label px-approved docs-approved

File: frontend/packages/console-dynamic-plugin-sdk/scripts/utils/typescript.ts (hunk 3 above)
Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/scripts/utils/typescript.ts]: I think this is one of those cases when TS thinks the value may be undefined but in practice it should never be undefined.

I'd just use https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html here to keep code simple:
```ts
relativePath(config.path!)
```

**Pull request merged**